lib/core/src/Cardano/Wallet/Primitive/Model.hs

```diff
@@ -269,20 +271,35 @@ updateState
     -> Wallet s
 updateState s (Wallet u tip _ bp) = Wallet u tip s bp
 
--- | Apply Block is the primary way of making the wallet evolve. It returns the
--- updated wallet state, as well as a set of all transactions belonging to the
--- wallet discovered while applying the block.
+-- | Represents a state transition that occurs when applying a single 'Block'
+--   to a 'Wallet' with the 'applyBlock' function.
+data WalletStateUpdate s = WalletStateUpdate
+    { initialState :: !(Wallet s)
+      -- ^ The initial state of the wallet.
+    , updatedState :: !(Wallet s)
+      -- ^ The updated state of the wallet.
+    , transactions :: ![(Tx, TxMeta)]
+      -- ^ Transactions that occur during the state transition.
+    , delegation :: !(Maybe PoolId)
+        -- ^ Any stake delegation that occurs during the state transition.
+    } deriving (Generic, Show, Eq)
+
+-- | Apply a single block to the wallet. This is the primary way of making a
+-- wallet evolve. It returns an updated wallet state, which includes a set of
+-- all transactions belonging to the wallet discovered while applying the block.
 applyBlock
-    :: Block
-    -> Wallet s
-    -> ([(Tx, TxMeta)], Wallet s)
-applyBlock !b (Wallet !u _ s bp) =
-    let
-        ((txs, u'), s') = prefilterBlock b u s
-    in
-        ( txs
-        , Wallet u' (b ^. #header) s' bp
-        )
+    :: Wallet s
+    -> Block
+    -> WalletStateUpdate s
```

Comment:
Introducing a type helper for the result of this function makes sense with the addition of the delegation certificate but this type has _too many responsibility_ and obfuscate a very idiomatic Haskell pattern (`s -> (result, s)`). 
Furthermore, the `initialState` sounds superfluous, this is already given as argument and, because of immutability, is already in-scope and unchanged at call sites.

I'd highly recommend the following as a middle-ground, preserving the state pattern while still allowing for a more readable output:
```hs
:: Wallet s
-> Block 
-> (ApplyBlockResult, Wallet s)
```
where:

```hs
 data ApplyBlockResult = ApplyBlockResult
    { ourTransactions :: ![(Tx, TxMeta)]
    , ourCertificates :: ![(PoolId, ChimericAccount)] -- Or alternatively ![PoolId]
    }
```

Also, delegation should really be a `[(PoolId, ChimericAccount)]` since there can be many certificates present in a single block :+1:

Reply:
@KtorZ wrote:
> I'd highly recommend the following as a middle-ground, preserving the state pattern while still allowing for a more readable output:
> ```hs
> :: Wallet s
> -> Block 
> -> (ApplyBlockResult, Wallet s)
> ```
Fair enough! I've updated the types to reflect this recommendation.

As for:
> Also, **delegation should really be a list** (e.g. [(PoolId, ChimericAccount)] or [PoolId]) since there can be many certificates present in a single block.

I checked and you're right that a _single block_ can encode _more than one_ delegation of stake from one account to _different pools_.

However, given this is the case, which (if any) of these delegations should take precedence? 

Moreover, our database (currently) only has space for one pool per `(wallet id, slot id)` combination:

```hs
DelegationCertificate 
    certWalletId             W.WalletId   sql=wallet_id   
    certSlot                 W.SlotId     sql=slot
    certPoolId               W.PoolId     sql=delegation   
```
Source: https://github.com/input-output-hk/cardano-wallet/blob/c2700b15a147ab8ec744d812ef483b9614597005/lib/core/src/Cardano/Wallet/DB/Sqlite/TH.hs#L146-L149

I've asked the Jörmungandr team to clarify this here: https://app.slack.com/client/T0N639Z4N/threads/thread/CCPKJ93Q9-1574305614.149500

Reply:
Further to previous comment.

The answer seems to be that the _**latest**_ delegation (in a block) for an account should take precedence over others.

Reply:
Of course :)
And that is what the database take care of already.

However, the internals of the primitive models shouldn't make such assumptions IMO and yield all certificates discovered in the block, in the same order as they're discovered. 

Yet that is debatable. One could argue that hiding this "detail" inside the primitive model would also make sense and prevent the result to be misinterpreted. I have no strong arguments that would favor one over the other.

Reply:
@KtorZ wrote:
> One could argue that hiding this "detail" inside the primitive model would also make sense and prevent the result to be misinterpreted. I have no strong arguments that would favor one over the other.

Assuming that `applyBlock` yields all delegation certificates discovered in a block, the caller will probably still need to filter out all but the last.

Reply:
> I have no strong arguments that would favor one over the other.

If the same block contains multiple delegations for the same wallet to different pools, I think there's a strong argument for writing this information to the log (rather than silently dropping every delegation except the last).

Since we have given `restoreBlocks` the responsibility of logging relevant data discovered in each block, I think it makes sense here to return _**all**_ delegations discovered in the `ApplyBlockResult` type. This way `restoreBlocks` can log each of them.